nri_device_injector/nri_device_injector_test.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2023 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParseDevices(t *testing.T) {
+	tests := map[string]struct {
+		container   string
+		annotations map[string]string
+		want        []device
+		wantErr     bool
+	}{
+		"No device annotations": {
+			container: "test",
+		},
+		"Unrelated device annotation": {
+			container:   "test",
+			annotations: map[string]string{"foo": "foo1"},
+		},
+		"One valid device annotation injecting to container": {
+			container: "test",
+			annotations: map[string]string{
+				"devices.gke.io/container.test": `
+- path: /dev/test
+`}, want: []device{{
+				Path: "/dev/test",
+			}},
+		},
+		"Multiple valid device annotation injecting to container": {
+			container: "test",
+			annotations: map[string]string{
+				"devices.gke.io/container.test": `
+- path: /dev/test0
+- path: /dev/test1
+  major: 123
+  minor: 456
+- path: /dev/test2
+  Type: b
+`}, want: []device{{
+				Path: "/dev/test0",
+			}, {
+				Path:  "/dev/test1",
+				Major: 123,
+				Minor: 456,
+			}, {
+				Path: "/dev/test2",
+				Type: "b",
+			}},
+		},
+		"Invalid device annotation": {
```

Comment:
Can we add a couple more test cases:
* empty annotation value
* invalid yaml
* devices defined multiple times

Reply:
Added. For the devices defined multiple times case, though I added a test case, I didn't really make the logic of getDevice to make sure there's only one annotation for each device path. And when a pod annotates a device multiple time, it's seen as there are multiple nri plugins trying to claim the same device and then the pod gets error creating container (https://github.com/containerd/nri/blob/main/pkg/adaptation/result.go#L1030). Or do we actually want to make sure each device's annotated at most once?

Reply:
> Or do we actually want to make sure each device's annotated at most once?

Yeah, I think it'd be better to have the plugin return an error if the same device is defined multiple times.

I'm fine with that being done in a follow-up PR rather than this one, so don't consider it a blocker for merging.

Reply:
Thanks Sam, I'll create another PR for this.